src/cogent3/core/annotation_db.py

```diff
@@ -37,13 +39,33 @@
 # Define custom types for storage in sqlite
 # https://stackoverflow.com/questions/18621513/python-insert-numpy-array-into-sqlite3-database
 def array_to_sqlite(data):
-    return data.tobytes()
+    out = io.BytesIO()
+    numpy.save(out, data)
+    out.seek(0)
+    return out.read()
 
 
 def sqlite_to_array(data):
-    result = numpy.frombuffer(data, dtype=int)
-    dim = result.shape[0] // 2
-    return result.reshape((dim, 2))
+    out = io.BytesIO(data)
+    out.seek(0)
+    try:
+        result = numpy.load(out)
+    except ValueError:
+        # array is not stored in the numpy.save format
+        # attempt to read from the old format where the
+        # array was saved using numpy.ndarray.tobytes
+        result = numpy.frombuffer(data, dtype=int)
+        dim = result.shape[0] // 2
+        result = result.reshape((dim, 2))
+
+        warnings.warn(
+            "Old OS dependent file format detected. "
+            "Update the file format using cogent3.core.annotation_db.update_file_format() "
+            "For reason see https://github.com/cogent3/cogent3/issues/1776.",
```

Comment:
I've been scratching my head for the last little bit on this. When this code is run through a WIP test I have going on in pytest, pytest is able to capture and display the warning. However, if I take the equivalent code in pytest and run it in its own file, the warnings are not displayed - unless the warning is also specifically given the parameter stacklevel=4.

Reply:
It does work if I use a `UserWarning` instead of a `DeprecationWarning`, so perhaps something weird going on with the filters.

Reply:
I'm also not entirely convinced this is the correct place to put the warning. For instance, the warning contains no information about the name of the bad file and would only be raised lazily when they attempt to do something that involves reading from the spans column. I think ideally, users should be warned when they load the database, but this would involve verifying all rows in every array type column on load.

Reply:
This is difficult. As you point out, doing it properly is a significant effort. It's better to have some warning rather than none. Let's leave it at that for now.